### PR TITLE
`ruff server`: Support setting to prioritize project configuration over editor configuration

### PR DESCRIPTION
**Pull request by snowsignal**
## Summary

This is intended to address https://github.com/astral-sh/ruff-vscode/issues/425, and is a follow-up to https://github.com/astral-sh/ruff/pull/11062.

A new client setting is now supported by the server, `prioritizeFileConfiguration`. This is a boolean setting (default: `false`) that, if set to `true`, will instruct the configuration resolver to prioritize file configuration (aka discovered TOML files) over configuration passed in by the editor.

A corresponding extension PR has been opened, which makes this setting available for VS Code: https://github.com/astral-sh/ruff-vscode/pull/457.

## Test Plan

To test this with VS Code, you'll need to check out [the VS Code PR](https://github.com/astral-sh/ruff-vscode/pull/457) that adds this setting.

The test process is similar to https://github.com/astral-sh/ruff/pull/11062, but in scenarios where the editor configuration would take priority over file configuration, file configuration should take priority.

### CODE REVIEW

File: crates/ruff_server/src/server/api/requests/code_action.rs [hunk 1]
```diff
@@ -140,7 +140,7 @@ fn organize_imports(snapshot: &DocumentSnapshot) -> crate::Result<CodeActionOrCo
                 document,
                 snapshot.resolved_client_capabilities(),
                 snapshot.url(),
-                &snapshot.settings().linter,
+                snapshot.settings().linter(),
```
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/code_action.rs:143]: These getters are perhaps a nice example of something that could be put up as a trivial standalone PR that would be immediately approved and merged :)

**Review by charliermarsh — COMMENTED** [review]
I wonder if this should be an enum, rather than a bool? That way, we can associate documentation with each variant too, which seems like it would be helpful.

**Comment by snowsignal** [comment]
> I wonder if this should be an enum, rather than a bool? That way, we can associate documentation with each variant too, which seems like it would be helpful.

@MichaReiser made a similar comment [here](https://github.com/astral-sh/ruff-vscode/pull/457#discussion_r1574236766). I think that's what I plan to go with :)

**Comment by charliermarsh** [comment]
Oops, I swear I didn't see that!

**Comment by snowsignal** [comment]
No worries 😄

File: crates/ruff_server/src/server/api/requests/code_action.rs (hunk 1 above)
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/code_action.rs:143]: Thanks, I'll keep that in mind for the future 👍

File: crates/ruff_server/src/session/settings.rs [hunk 2]
```diff
@@ -38,6 +38,22 @@ pub(crate) struct ResolvedEditorSettings {
...
+/// Determines how multiple conflicting configurations should be resolved - in this
+/// case, the configuration from the client settings and configuration from local
+/// `.toml` files (aka 'workspace' configuration).
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+pub(crate) enum ConfigResolutionStrategy {
+    /// The default strategy - configuration set in the editor takes priority over
+    /// workspace configuration set in `.toml` files.
+    Default,
```
Comment (charliermarsh) [inline crates/ruff_server/src/session/settings.rs:53]: I think we should have a name here that describes the strategy, rather than using `Default`.

File: crates/ruff_server/src/session/settings.rs [hunk 3]
```diff
@@ -38,6 +38,22 @@ pub(crate) struct ResolvedEditorSettings {
     pub(super) ignore: Option<Vec<RuleSelector>>,
     pub(super) exclude: Option<Vec<String>>,
     pub(super) line_length: Option<LineLength>,
+    pub(super) configuration_resolution_strategy: ConfigResolutionStrategy,
```
Comment (charliermarsh) [inline crates/ruff_server/src/session/settings.rs:41]: I'd like to find a way to make the name both shorter and more obvious to users. "Configuration preference"?

File: crates/ruff_server/src/session/settings.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff_server/src/session/settings.rs:53]: Could these be "Editor" and "Workspace" (or "Filesystem"), if the setting were called "configurationPreference"?

Comment (charliermarsh) [reply crates/ruff_server/src/session/settings.rs:53]: (I'd also suggest making it `#[default]`.)

**Review by charliermarsh — COMMENTED** [review]
The code looks good, I'm just not sold on the names...

**Comment by charliermarsh** [comment]
I think @MichaReiser will have good advice on the naming and documentation.

**Comment by MichaReiser** [comment]
> I think @MichaReiser will have good advice on the naming and documentation.

Lol no :laughing: I commented in the other PR that I don't have good naming suggestions haha

File: crates/ruff_server/src/session/settings.rs [hunk 4]
```diff
@@ -38,6 +38,22 @@ pub(crate) struct ResolvedEditorSettings {
...
+}
+
+/// Determines how multiple conflicting configurations should be resolved - in this
+/// case, the configuration from the client settings and configuration from local
+/// `.toml` files (aka 'workspace' configuration).
+#[derive(Clone, Copy, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+pub(crate) enum ConfigResolutionStrategy {
```
Comment (MichaReiser) [inline crates/ruff_server/src/session/settings.rs:50]: Nit: `Config` -> `Configuration` for consistency (you use configuration in most other places)

File: crates/ruff_server/src/session/settings.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/ruff_server/src/session/settings.rs:53]: I like Charlie's suggestion. My first thought was `PrioritizeEditor`. I think I would not call it resolution, because I associate resolution with finding the configuration. Althought that's probably just me.

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_server/src/session/settings.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff_server/src/session/settings.rs:53]: Yeah I think I'd be fine approving either `ConfigurationPreference` or `ConfigurationPrecedence`, with options like:

- `Editor`
- `Workspace`

Alternatively, if we _do_ want an option that uses the editor and _ignores_ workspace configuration (which is what I initially expected it to do), we could have:

- `EditorFirst` (equivalent to the `Default` here)
- `FilesystemFirst` (equivalent to `PrioritizeWorkspace` here)
- `EditorOnly` (use editor settings, ignore on-disk configuration)

Comment (snowsignal) [reply crates/ruff_server/src/session/settings.rs:53]: I like `ConfigurationPreference` and `EditorFirst`/`FilesystemFirst`/`EditorOnly` 👍
